File: dev/requirements/behavior/cases/library-survives-relaunch.1.6.case.swift
import Foundation
import SetlistCore

/// 1.6 — The library survives app relaunch: jokes, sets and shows written in
/// one session are read back in the next.
let librarySurvivesRelaunch_1_6 = RequirementCase(id: "1.6") {
    let container = FileManager.default.temporaryDirectory
        .appendingPathComponent("setlist-1.6-\(UUID().uuidString)")
    defer { try? FileManager.default.removeItem(at: container) }

    let joke = Joke(title: "Airport security", body: "…", estimatedLength: 150)
    let set = SetList(name: "Fri late", jokeIDs: [joke.id])
    let show = Show(
        venue: "The Stand",
        startedAt: REFERENCE_NOW,
        plannedLength: 1200,
        audience: Audience(size: 80, seating: .standing),
        setListID: set.id
    )
    let library = Library(jokes: [joke], sets: [set], shows: [show])

    try FileLibraryStore.inContainer(container).save(library)

    // A second store over the same container is the next launch: nothing is
    // carried in memory between them.
    let relaunched = try FileLibraryStore.inContainer(container).load()
    try expectEqual(relaunched, library, "the library read back after a relaunch")
    try expectEqual(relaunched.jokes(of: relaunched.sets[0]).map(\.title), ["Airport security"], "the set still resolves")

    // A first launch, with nothing written yet, is an empty library rather than
    // an error the comedian has to dismiss before their set.
    let fresh = FileManager.default.temporaryDirectory
        .appendingPathComponent("setlist-1.6-fresh-\(UUID().uuidString)")
    defer { try? FileManager.default.removeItem(at: fresh) }
    try expectEqual(try FileLibraryStore.inContainer(fresh).load(), Library(), "a first launch")
}

File: dev/requirements/behavior/cases/library-survives-relaunch.1.6.case.dart
import 'dart:io';

import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 1.6 — The library survives app relaunch: jokes, sets and shows written in
/// one session are read back in the next.
final librarySurvivesRelaunch_1_6 = RequirementCase('1.6', (tester) async {
  // Real file I/O only completes outside the fake-async test zone: inside it,
  // the future never resolves and the case hangs until its timeout.
  await tester.runAsync(() async {
    final container = await Directory.systemTemp.createTemp('setlist-relaunch');
    try {
    await FileLibraryStore.inContainer(container).save(
      Library(
        jokes: [
          Joke(
            id: 'airport',
            title: 'Airport security',
            body: 'Shoes off.',
            estimatedLength: const Duration(seconds: 145),
          ),
        ],
        sets: [SetList(id: 'fri', name: 'Fri late', jokeIds: ['airport'])],
        shows: [
          Show(
            id: 'stand',
            venue: 'The Stand',
            startedAt: DateTime.utc(2026, 8, 21, 22, 30),
            plannedLength: const Duration(minutes: 20),
            audience: const Audience(size: 80, seating: Seating.seated),
          ),
        ],
      ),
    );

    // A second store over the same container is the next launch.
    final read = await FileLibraryStore.inContainer(container).load();

    expectThat(read.jokes.single.title, 'Airport security', 'the material is still there');
    expectThat(
      read.jokes.single.estimatedLength,
      const Duration(seconds: 145),
      'with its estimate intact',
    );
    expectThat(read.sets.single.jokeIds, ['airport'], 'the set still names its jokes');
    expectThat(read.shows.single.venue, 'The Stand', 'and the show still knows the room');
    expectThat(read.shows.single.audience.size, 80, 'including who was in it');
    } finally {
      await container.delete(recursive: true);
    }
  });
});
